src/atomWithSearchParams.ts

```diff
@@ -0,0 +1,151 @@
+import type { SetStateAction, WritableAtom } from 'jotai/vanilla';
+import { atom } from 'jotai/vanilla';
+import { atomWithLocation } from './atomWithLocation';
+
+// Create an atom for managing location state, including search parameters.
+const locationAtom = atomWithLocation();
+
+/**
+ * Creates an atom that manages a single search parameter.
+ *
+ * The atom automatically infers the type of the search parameter based on the
+ * type of `defaultValue`.
+ *
+ * The atom's read function returns the current value of the search parameter.
+ * The atom's write function updates the search parameter in the URL.
+ *
+ * @param key - The key of the search parameter.
+ * @param defaultValue - The default value of the search parameter.
+ * @returns A writable atom that manages the search parameter.
+ */
+export const atomWithSearchParams = <T>(
+  key: string,
+  defaultValue: T,
+): WritableAtom<T, [SetStateAction<T>], void> => {
+  /**
+   * Resolves the value of a search parameter based on the type of `defaultValue`.
+   *
+   * @param value - The raw value from the URL (could be `null` or `undefined`).
+   * @returns The resolved value matching the type of `defaultValue`.
+   */
+  const resolveValue = (value: string | null | undefined): T => {
+    // If the value is null, undefined, or not a string, return the default value.
+    if (value === null || value === undefined || typeof value !== 'string') {
```

Comment:
Just a preference, but let's do this:
```suggestion
    if (typeof value !== 'string') {
```